functions/Get-DbaRegisteredServer.ps1

```diff
@@ -0,0 +1,164 @@
+
+			Returns a list of servers in the HR and sub-group Development from the CMS on sqlserver2014a
+	#>
+	[CmdletBinding(DefaultParameterSetName = "Default")]
+	param (
+		[parameter(Mandatory = $true, ValueFromPipeline = $true)]
+		[Alias("ServerInstance", "SqlServer")]
+		[DbaInstanceParameter[]]$SqlInstance,
+		[PSCredential]$SqlCredential,
+		[Alias("Groups")]
+		[object[]]$Group,
+		[object[]]$ExcludeGroup,
+		[switch]$ExcludeCmsServer,
+		[switch]$NameOnly,
+		[switch]$Silent
+	)
+	begin {
+		function Find-CmsGroup {
+			[cmdletbinding()]
+			param(
+				$CmsGrp,
+				$Base = $null,
+				$Stopat
+			)
+			$results = @()
+			foreach ($el in $CmsGrp) {
+				if ($null -eq $Base -or [string]::IsNullOrWhiteSpace($Base) ) {
+					$partial = $el.name
+				}
+				else {
+					$partial = "$Base\$($el.name)"
+				}
+				if ($partial -eq $Stopat) {
+					return $el
+				}
+				else {
+					foreach ($elg in $el.ServerGroups) {
+						$results += Find-CmsGroup -CmsGrp $elg -Base $partial -Stopat $Stopat
+					}
+				}
+			}
+			return $results
+		}
+	}
+
+	process {
+		if (Test-FunctionInterrupt) { return }
+
+		$servers = @()
+		foreach ($instance in $SqlInstance) {
+			try {
+				$cmsStore = Get-DbaRegisteredServersStore -SqlInstance $instance -SqlCredential $SqlCredential -Silent:$Silent
+			}
+			catch {
+				Stop-Function -Message "Cannot access Central Management Server '$instance'" -ErrorRecord $_ -Continue
+				return
+			}
+
+			if (Test-Bound -ParameterName ExcludeGroup) {
+				$Group = ($cmsStore.DatabaseEngineServerGroup.ServerGroups | Where-Object Name -notin $ExcludeGroup).Name
+			}
+
+			if ($Group) {
+				foreach ($currentGroup in $Group) {
+					$cms = Find-CmsGroup -CmsGrp $cmsStore.DatabaseEngineServerGroup.ServerGroups -Stopat $currentGroup
+					if ($null -eq $cms) {
+						Write-Message -Level Output -Message "No groups found matching that name on instance '$instance'"
+						continue
+					}
+					$servers += ($cms.GetDescendantRegisteredServers())
+				}
+			}
+			else {
+				$cms = $cmsStore.DatabaseEngineServerGroup
+				$servers += ($cms.GetDescendantRegisteredServers())
+			}
+
+			if ($ExcludeCmsServer) {
+				$servers = ($servers | Where-Object { $_.ServerName -ne $instance})
+			}
+		}
+	}
+
+	end {
+		foreach ($server in $servers) {
+			try {
+				$lookup = Resolve-DbaNetworkName $server.ServerName -Turbo -Silent
+				Add-Member -Force -InputObject $server -MemberType NoteProperty -Name ComputerName -Value $lookup.ComputerName
+				Add-Member -Force -InputObject $server -MemberType NoteProperty -Name IPAddress -Value $lookup.IPAddress
+			}
+			catch {
+				Add-Member -Force -InputObject $server -MemberType NoteProperty -Name ComputerName -Value $null
+				Add-Member -Force -InputObject $server -MemberType NoteProperty -Name IPAddress -Value $null
+			}
+		}
+
+		if ($NameOnly) {
+			return $servers | Select-Object -ExpandProperty ServerName -Unique
+		}
+
+		return $servers
```

Comment:
Results should be returned as early as possible.
All steps can be executed in the `process` instead of in the `end` block.